temporalcli/commands.workflow.go

```diff
@@ -252,3 +329,147 @@ func startBatchJob(cctx *CommandContext, cl client.Client, req *workflowservice.
+		return getLastContinueAsNewID(ctx, namespace, wid, rid, wfsvc)
+	case "FirstWorkflowTask":
+		return getFirstWorkflowTaskEventID(ctx, namespace, wid, rid, wfsvc)
+	default:
+		return "", -1, fmt.Errorf("invalid reset type: %s", resetType)
+	}
+}
+
+// Returns event id of the last completed task or id of the next event after scheduled task.
+func getLastWorkflowTaskEventID(ctx context.Context, namespace, wid, rid string, wfsvc workflowservice.WorkflowServiceClient) (resetBaseRunID string, workflowTaskEventID int64, err error) {
+	resetBaseRunID = rid
+	req := workflowservice.GetWorkflowExecutionHistoryReverseRequest{
+		Namespace: namespace,
+		Execution: &commonpb.WorkflowExecution{
+			WorkflowId: wid,
+			RunId:      rid,
+		},
+		MaximumPageSize: 250,
+		NextPageToken:   nil,
+	}
+
+	for {
+		resp, err := wfsvc.GetWorkflowExecutionHistoryReverse(ctx, &req)
+		if err != nil {
+			return "", 0, fmt.Errorf("failed to get workflow execution history: %w", err)
+		}
+		for _, e := range resp.GetHistory().GetEvents() {
+			if e.GetEventType() == enumspb.EVENT_TYPE_WORKFLOW_TASK_COMPLETED {
+				workflowTaskEventID = e.GetEventId()
+				break
+			} else if e.GetEventType() == enumspb.EVENT_TYPE_WORKFLOW_TASK_SCHEDULED {
+				// if there is no task completed event, set it to first scheduled event + 1
+				workflowTaskEventID = e.GetEventId() + 1
+			}
+		}
+		if len(resp.NextPageToken) != 0 {
+			req.NextPageToken = resp.NextPageToken
+		} else {
+			break
+		}
+	}
+	if workflowTaskEventID == 0 {
+		return "", 0, errors.New("unable to find any scheduled or completed task")
+	}
+	return
+}
+
+// Returns id of the first workflow task completed event or if it doesn't exist then id of the event after task scheduled event.
+func getFirstWorkflowTaskEventID(ctx context.Context, namespace, wid, rid string, wfsvc workflowservice.WorkflowServiceClient) (resetBaseRunID string, workflowTaskEventID int64, err error) {
+	resetBaseRunID = rid
+	req := workflowservice.GetWorkflowExecutionHistoryRequest{
+		Namespace: namespace,
+		Execution: &commonpb.WorkflowExecution{
+			WorkflowId: wid,
+			RunId:      rid,
+		},
+		MaximumPageSize: 250,
+		NextPageToken:   nil,
+	}
+	for {
+		resp, err := wfsvc.GetWorkflowExecutionHistory(ctx, &req)
+		if err != nil {
+			return "", 0, fmt.Errorf("failed to get workflow execution history: %w", err)
+		}
+		for _, e := range resp.GetHistory().GetEvents() {
+			if e.GetEventType() == enumspb.EVENT_TYPE_WORKFLOW_TASK_COMPLETED {
+				workflowTaskEventID = e.GetEventId()
+				return resetBaseRunID, workflowTaskEventID, nil
+			}
+			if e.GetEventType() == enumspb.EVENT_TYPE_WORKFLOW_TASK_SCHEDULED {
+				if workflowTaskEventID == 0 {
+					workflowTaskEventID = e.GetEventId() + 1
+				}
+			}
+		}
+		if len(resp.NextPageToken) != 0 {
+			req.NextPageToken = resp.NextPageToken
+		} else {
+			break
+		}
+	}
+	if workflowTaskEventID == 0 {
+		return "", 0, errors.New("unable to find any scheduled or completed task")
+	}
```

Comment:
I think this part is the same as in the previous function.

Reply:
It is except one iterates forwards through history and the other in reverse order, see `GetWorkflowExecutionHistory` vs `GetWorkflowExecutionHistoryReverse`.

The logic in the `EVENT_TYPE_WORKFLOW_TASK_SCHEDULED` case also differs